rfc/rfc-0002-bindings.md

```diff
@@ -0,0 +1,95 @@
+# RFC-0002: Bindings
+
+**To be Reviewed By:** 
+
+**Authors:** @scothis
+
+**Status:** 
+
+**Pull Request URL:** [#1360](https://github.com/projectriff/riff/pull/1360)
+
+**Superseded by:** N/A
+
+**Supersedes:** N/A
+
+**Related:** N/A
+
+
+## Problem
+Non-trivial workloads often need to access resources, like databases, message brokers or any credentialed resource. In Cloud Foundry, these are called services and they are provisioned from a broker and then bound to applications at both build and runtime. As Kubernetes has a different notion of a Service, this RFC will refer to "service bindings" as simply bindings.
+
+The binding contains both metadata and credentials to provide a workload the details it needs to connect and consume the resource.
+
+### Anti-Goals
+A generic mechanism to provision bindings, and the specific structure of the metadata and credentials are out of scope for this RFC.
+
+## Solution
+The Cloud Native Buildpacks project has a [spec for Bindings](https://github.com/buildpack/spec/blob/master/extensions/bindings.md) which defines the form bindings should take at both build and runtime. In short, each binding is represented as a directory inside a well known directory with two sub-directories: `metadata` for general non-sensitive information about the binding, and `secret` for credentials and other sensitive information. Each attribute for the binding is a file in one of those two directories. The metadata is available at buildtime and both the metadata and secret is available at runtime.
+
+riff should adopt CNB Bindings spec for bindings at both build and runtime.
+
+Kubernetes uses [Volumes](https://kubernetes.io/docs/concepts/storage/volumes/) to inject files from an external source into a running container. While there is a diverse spectrum of volume implementations available, in order to maximize both compatibility with bog standard Kubernetes and simplify the implementation, we should focus on a [configMap volume](https://kubernetes.io/docs/concepts/storage/volumes/#configmap) for binding metadata and a [secret volume](https://kubernetes.io/docs/concepts/storage/volumes/#secret) for binding secrets, which map directly to Kubernetes [ConfigMaps](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/) and [Secrets](https://kubernetes.io/docs/concepts/configuration/secret/) respectively.
+
+As a binding is backed by a ConfigMap and Secret a mechanism is needed to specify which ConfigMap and Secret. [LocalObjectReferences](https://godoc.org/k8s.io/api/core/v1#LocalObjectReference) are commonly used to define the coordinates to an arbitrary resources. A binding reference consists of `metadata` LocalObjectReference to a ConfigMap and a `secret` LocalObjectReference to a Secret.
+
+Any resource that provides a binding should include the binding reference on its status, for example:
```

Comment:
Who has the responsibility for making the resource adhere to this requirement? I do not see us doing it, which means it will be an additional step that our users will have to undertake.

It would be useful to list the steps that a user has to when when they have provisioned say a mysql using a helm chart, and wants to bind to it their function.

Reply:
It's encouraged for resources to publicize the binding reference, but is not required. It's an anti-goal of this RFC to define such a mechanism.

As there is no requirement, there is no enforcement mechanism. If desired, such a requirement/mechanism should be the focus of another RFC.